File: Frontend-BAP/screens/ProviderItemsScreen.js
import React, { useState } from "react";
import axios from 'axios';
import {
  View,
  Text,
  Image,
  TouchableOpacity,
  ScrollView,
  KeyboardAvoidingView,
  Platform,
  Alert,
} from "react-native";
import tw from "tailwind-react-native-classnames";
import { SERVER_URL,mobile_url } from '@env';

export default function ProviderItemsScreen({ navigation, route }) {
  const [provider, setProvider] = useState(route.params?.provider);
  const isOrganic=true;
  console.log(provider);
 const handleInit = async () => {
  try {
    const response = await axios.post(
  `${SERVER_URL}/bap/init`,
      {
        provider_id: provider.provider_id, // ✅ Fix: use provider_id instead of provider.id
        items: provider.items.map(item => ({
          id: item.bpp_product_id, // ✅ Use correct key
          quantity: item.quantity,
          fulfillment_id: item.fulfillment_id
        })),
        delivery_address: {
          gps: "23.0225,72.5714",
          address: "Plot 12, SG Road, Ahmedabad, Gujarat"
        }
      },
      {
        headers: {
          'x-transaction-id': 'txn-' + Date.now()
        }
      }
    );

    const bpp_response = response.data.bpp_response;
    // console.log("✅ BPP Response:", bpp_response);
    console.log("running");
    
    navigation.navigate("VerifyProductsScreen", { bpp_response });

  } catch (error) {
    console.error("❌ Init API failed:", error.message);
    Alert.alert("Failed to verify products. Please try again.");
  }
};



  
  const updateCartQuantity = async (item, newQty) => {
  try {
    const response = await fetch(`${SERVER_URL}/cart/update`, {
      method: "PUT", // ✅ Use PUT
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        user_id: "a985baac-9028-4dc1-bbd9-a6f3aae49ef5",
        bpp_product_id: item.bpp_product_id,
        provider_id: provider.provider_id,
        quantity: newQty,
      }),
    });

    // Check for proper content type
    const contentType = response.headers.get("content-type");
    if (!response.ok) {
      throw new Error(`HTTP ${response.status} - ${response.statusText}`);
    }

    if (contentType && contentType.includes("application/json")) {
      const data = await response.json();
      console.log("✅ Quantity updated", data);

      let updatedItems;

      if (newQty === 0) {
        // ❌ Remove item from list if quantity is 0
        updatedItems = provider.items.filter(
          (i) => i.bpp_product_id !== item.bpp_product_id
        );
      } else {
        // ✅ Update item's quantity and total_price
        updatedItems = provider.items.map((i) =>
          i.bpp_product_id === item.bpp_product_id
            ? {
                ...i,
                quantity: newQty,
                total_price: parseFloat((newQty * (i.unit_price || 0)).toFixed(2)),
              }
            : i
        );
      }
      setProvider({ ...provider, items: updatedItems });
    } else {
      const html = await response.text(); // catch HTML error responses
      console.error("❌ Server returned HTML instead of JSON:", html);
    }
  } catch (err) {
    console.error("⚠️ Error updating quantity", err.message || err);
  }
};


  if (!provider) {
    return (
      <View style={tw`flex-1 justify-center items-center`}>
        <Text style={tw`text-red-600 text-lg font-semibold`}>
          ⚠️ No provider data available
        </Text>
      </View>
    );
  }

  const totalMRP = (provider.items || []).reduce(
    (sum, item) => sum + (item.total_price || 0),
    0
  );

return (
  <KeyboardAvoidingView
    style={tw`flex-1 bg-green-50`}
    behavior={Platform.OS === "ios" ? "padding" : "height"}
    keyboardVerticalOffset={Platform.OS === "ios" ? 100 : 0}
  >
    <ScrollView
      style={tw`p-4`}
      contentContainerStyle={{ paddingBottom: 160 }}
      showsVerticalScrollIndicator={false}
    >
      {/* 🏪 Provider Header */}
      <View style={tw`bg-green-100 p-5 rounded-2xl mb-6 shadow-md`}>
        <Text style={tw`text-2xl font-extrabold text-green-900`}>
          {provider.provider_name}
        </Text>
        <Text style={tw`text-sm text-green-700 mt-1`}>
          📍 {provider.provider_address}
        </Text>
      </View>

      {/* 🥕 Product List */}
      {(provider.items || []).map((item, index) => {
        const isOrganic = item.tags?.some(t => t.key === "organic" && t.value === "true");

        return (
          <TouchableOpacity
            key={index}
            activeOpacity={0.9}
            style={tw`bg-white p-4 mb-5 rounded-2xl border border-green-100 shadow-md`}
            onPress={() =>
              navigation.navigate("ProductDetails", { item_id: item.bpp_product_id })
            }
          >
            <View style={tw`flex-row items-center`}>
              {/* Thumbnail */}
              <View style={tw`relative mr-4`}>
                <Image
                  source={{ uri: item.image_url || "https://via.placeholder.com/60" }}
                  style={tw`w-20 h-20 rounded-xl`}
                  resizeMode="cover"
                />
                {isOrganic && (
                  <View
                    style={tw`absolute -top-2 -right-2 bg-green-200 px-2 py-1 rounded-full shadow-sm`}
                  >
                    <Text style={tw`text-green-800 text-xs font-semibold`}>
                      Organic
                    </Text>
                  </View>
                )}
              </View>

              {/* Product Info */}
              <View style={tw`flex-1`}>
                <Text style={tw`font-bold text-gray-900 text-lg`} numberOfLines={1}>
                  {item.item_name}
                </Text>

                {/* Quantity Controls */}
                <View style={tw`flex-row items-center mt-3`}>
                  <TouchableOpacity
                    onPress={() => updateCartQuantity(item, item.quantity - 1)}
                    style={tw`w-9 h-9 rounded-full bg-gray-200 justify-center items-center shadow mr-3`}
                    activeOpacity={0.7}
                  >
                    <Text style={tw`text-xl font-bold text-gray-700`}>−</Text>
                  </TouchableOpacity>

                  <Text style={tw`text-lg font-bold text-gray-800 mx-2`}>
                    {item.quantity}
                  </Text>

                  <TouchableOpacity
                    onPress={() => updateCartQuantity(item, item.quantity + 1)}
                    style={tw`w-9 h-9 rounded-full bg-gray-200 justify-center items-center shadow ml-3`}
                    activeOpacity={0.7}
                  >
                    <Text style={tw`text-xl font-bold text-gray-700`}>+</Text>
                  </TouchableOpacity>
                </View>

                {/* Price Row */}
                <View style={tw`flex-row justify-between mt-2 items-center`}>
                  <Text style={tw`text-sm text-green-700`}>
                    Unit Price: ₹{item.unit_price}
                  </Text>
                  <Text style={tw`text-green-900 font-extrabold text-lg`}>
                    ₹{item.total_price}
                  </Text>
                </View>
              </View>
            </View>
          </TouchableOpacity>
        );
      })}
    </ScrollView>

    {/* 📌 Sticky Footer */}
    <View
      style={[
        tw`absolute bottom-0 left-0 right-0 bg-white border-t border-gray-200 px-5 py-4`,
        { shadowColor: "#000", shadowOpacity: 0.08, shadowRadius: 8, elevation: 10 },
      ]}
    >
      {/* Total Summary */}
      <View style={tw`bg-green-50 p-4 rounded-xl shadow-sm mb-4`}>
        <Text style={tw`font-bold text-gray-800 text-lg`}>
          Total Amount: ₹{totalMRP + 40}
        </Text>
        <Text style={tw`text-xs text-gray-600 mt-1`}>
          Includes ₹40 shipping
        </Text>
      </View>

      {/* CTA */}
      <TouchableOpacity
        onPress={handleInit}
        activeOpacity={0.9}
        style={tw`bg-green-700 py-4 rounded-full shadow-lg`}
      >
        <Text style={tw`text-white font-bold text-center text-lg`}>
          ✅ Verify Products & Continue
        </Text>
      </TouchableOpacity>
    </View>
  </KeyboardAvoidingView>
);


}

File: demo/lib/screens/provider_items_screen.dart
import 'dart:convert';

import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;

class ProviderItemsScreen extends StatefulWidget {
  final Map<String, dynamic> provider;
  const ProviderItemsScreen({Key? key, required this.provider})
      : super(key: key);

  @override
  State<ProviderItemsScreen> createState() => _ProviderItemsScreenState();
}

class _ProviderItemsScreenState extends State<ProviderItemsScreen> {
  late Map<String, dynamic> provider;
  final shippingCost = 40;

  @override
  void initState() {
    super.initState();
    provider = widget.provider;
  }

  double get totalMRP {
    return (provider['items'] as List<dynamic>)
        .fold(0.0, (sum, item) => sum + (item['total_price'] ?? 0));
  }

  Future<void> handleInit() async {
    try {
      final response = await http.post(
        Uri.parse('YOUR_SERVER_URL/bap/init'),
        headers: {
          'Content-Type': 'application/json',
          'x-transaction-id': 'txn-${DateTime.now().millisecondsSinceEpoch}',
        },
        body: jsonEncode({
          'provider_id': provider['provider_id'],
          'items': (provider['items'] as List<dynamic>).map((item) {
            return {
              'id': item['bpp_product_id'],
              'quantity': item['quantity'],
              'fulfillment_id': item['fulfillment_id']
            };
          }).toList(),
          'delivery_address': {
            'gps': '23.0225,72.5714',
            'address': 'Plot 12, SG Road, Ahmedabad, Gujarat'
          }
        }),
      );

      if (response.statusCode == 200) {
        final bppResponse = jsonDecode(response.body)['bpp_response'];
        // Navigate to next screen (replace with your route)
        Navigator.pushNamed(context, '/verifyProducts',
            arguments: {'bpp_response': bppResponse});
      } else {
        throw Exception('Failed to verify products');
      }
    } catch (e) {
      print('Error: $e');
      ScaffoldMessenger.of(context).showSnackBar(
        const SnackBar(content: Text('Failed to verify products. Please try again.')),
      );
    }
  }

  void updateCartQuantity(Map<String, dynamic> item, int newQty) async {
    try {
      final response = await http.put(
        Uri.parse('YOUR_SERVER_URL/cart/update'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'user_id': 'a985baac-9028-4dc1-bbd9-a6f3aae49ef5',
          'bpp_product_id': item['bpp_product_id'],
          'provider_id': provider['provider_id'],
          'quantity': newQty
        }),
      );

      if (response.statusCode == 200) {
        List<dynamic> updatedItems;
        if (newQty == 0) {
          updatedItems = (provider['items'] as List<dynamic>)
              .where((i) => i['bpp_product_id'] != item['bpp_product_id'])
              .toList();
        } else {
          updatedItems = (provider['items'] as List<dynamic>).map((i) {
            if (i['bpp_product_id'] == item['bpp_product_id']) {
              i['quantity'] = newQty;
              i['total_price'] = double.parse(
                  (newQty * (i['unit_price'] ?? 0)).toStringAsFixed(2));
            }
            return i;
          }).toList();
        }
        setState(() {
          provider['items'] = updatedItems;
        });
      } else {
        print('Failed to update cart: ${response.body}');
      }
    } catch (e) {
      print('Error updating quantity: $e');
    }
  }

  @override
  Widget build(BuildContext context) {
    final items = provider['items'] as List<dynamic>;
    return Scaffold(
      backgroundColor: Colors.green[50],
      appBar: AppBar(
        title: Text(provider['provider_name']),
        backgroundColor: Colors.green,
      ),
      body: Stack(
        children: [
          Padding(
            padding: const EdgeInsets.only(bottom: 150),
            child: SingleChildScrollView(
              padding: const EdgeInsets.all(16),
              child: Column(
                children: [
                  // Provider Header
                  Container(
                    padding: const EdgeInsets.all(16),
                    decoration: BoxDecoration(
                      color: Colors.green[100],
                      borderRadius: BorderRadius.circular(20),
                    ),
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        Text(
                          provider['provider_name'],
                          style: const TextStyle(
                              fontSize: 24, fontWeight: FontWeight.bold, color: Colors.green),
                        ),
                        const SizedBox(height: 4),
                        Text(
                          "📍 ${provider['provider_address']}",
                          style: TextStyle(fontSize: 14, color: Colors.green[800]),
                        ),
                      ],
                    ),
                  ),
                  const SizedBox(height: 16),

                  // Product List
                  ...items.map((item) {
                    final isOrganic = (item['tags'] as List<dynamic>?)
                            ?.any((t) => t['key'] == 'organic' && t['value'] == 'true') ??
                        false;

                    return Container(
                      margin: const EdgeInsets.only(bottom: 16),
                      padding: const EdgeInsets.all(16),
                      decoration: BoxDecoration(
                        color: Colors.white,
                        borderRadius: BorderRadius.circular(20),
                        border: Border.all(color: Colors.green.shade100),
                      ),
                      child: Row(
                        crossAxisAlignment: CrossAxisAlignment.start,
                        children: [
                          // Thumbnail
                          Stack(
                            children: [
                              ClipRRect(
                                borderRadius: BorderRadius.circular(16),
                                child: Image.network(
                                  item['image_url'] ??
                                      'https://via.placeholder.com/60',
                                  width: 80,
                                  height: 80,
                                  fit: BoxFit.cover,
                                ),
                              ),
                              if (isOrganic)
                                Positioned(
                                  top: -4,
                                  right: -4,
                                  child: Container(
                                    padding: const EdgeInsets.symmetric(
                                        horizontal: 6, vertical: 2),
                                    decoration: BoxDecoration(
                                      color: Colors.green[200],
                                      borderRadius: BorderRadius.circular(12),
                                    ),
                                    child: const Text(
                                      'Organic',
                                      style: TextStyle(
                                          fontSize: 10,
                                          fontWeight: FontWeight.bold,
                                          color: Colors.green),
                                    ),
                                  ),
                                ),
                            ],
                          ),
                          const SizedBox(width: 12),

                          // Product Info
                          Expanded(
                            child: Column(
                              crossAxisAlignment: CrossAxisAlignment.start,
                              children: [
                                Text(
                                  item['item_name'],
                                  style: const TextStyle(
                                      fontSize: 16, fontWeight: FontWeight.bold),
                                  maxLines: 1,
                                  overflow: TextOverflow.ellipsis,
                                ),
                                const SizedBox(height: 8),

                                // Quantity Controls
                                Row(
                                  children: [
                                    IconButton(
                                      onPressed: () => updateCartQuantity(
                                          item, item['quantity'] - 1),
                                      icon: const Icon(Icons.remove_circle_outline),
                                    ),
                                    Text(
                                      '${item['quantity']}',
                                      style: const TextStyle(
                                          fontSize: 16, fontWeight: FontWeight.bold),
                                    ),
                                    IconButton(
                                      onPressed: () => updateCartQuantity(
                                          item, item['quantity'] + 1),
                                      icon: const Icon(Icons.add_circle_outline),
                                    ),
                                  ],
                                ),
                                const SizedBox(height: 8),

                                // Price Row
                                Row(
                                  mainAxisAlignment:
                                      MainAxisAlignment.spaceBetween,
                                  children: [
                                    Text('Unit Price: ₹${item['unit_price']}',
                                        style: TextStyle(
                                            color: Colors.green[700], fontSize: 12)),
                                    Text('₹${item['total_price']}',
                                        style: const TextStyle(
                                            fontWeight: FontWeight.bold, fontSize: 16)),
                                  ],
                                ),
                              ],
                            ),
                          )
                        ],
                      ),
                    );
                  }).toList(),
                ],
              ),
            ),
          ),

          // Sticky Footer
          Positioned(
            bottom: 0,
            left: 0,
            right: 0,
            child: Container(
              padding: const EdgeInsets.all(16),
              decoration: BoxDecoration(
                color: Colors.white,
                border: Border(top: BorderSide(color: Colors.grey.shade300)),
                boxShadow: [
                  BoxShadow(
                    color: Colors.black.withOpacity(0.08),
                    blurRadius: 8,
                  )
                ],
              ),
              child: Column(
                mainAxisSize: MainAxisSize.min,
                children: [
                  Container(
                    padding: const EdgeInsets.all(12),
                    decoration: BoxDecoration(
                        color: Colors.green[50],
                        borderRadius: BorderRadius.circular(12)),
                    child: Row(
                      mainAxisAlignment: MainAxisAlignment.spaceBetween,
                      children: [
                        const Text(
                          'Total Amount',
                          style: TextStyle(
                              fontWeight: FontWeight.bold, fontSize: 16),
                        ),
                        Text(
                          '₹${totalMRP + shippingCost}',
                          style: const TextStyle(
                              fontWeight: FontWeight.bold, fontSize: 16),
                        ),
                      ],
                    ),
                  ),
                  const SizedBox(height: 8),
                  ElevatedButton(
                    style: ElevatedButton.styleFrom(
                        backgroundColor: Colors.green,
                        padding: const EdgeInsets.symmetric(vertical: 16),
                        shape: RoundedRectangleBorder(
                            borderRadius: BorderRadius.circular(12))),
                    onPressed: handleInit,
                    child: const Center(
                        child: Text(
                      '✅ Verify Products & Continue',
                      style: TextStyle(
                          color: Colors.white, fontWeight: FontWeight.bold),
                    )),
                  )
                ],
              ),
            ),
          ),
        ],
      ),
    );
  }
}
